### ISSUE TITLE
No support for SystemNotifier tray icons (electron apps)

### ISSUE THREAD

**Issue by Kryszak** [polybar/polybar#2010]
## Is your feature request related to a problem? Please describe.
Some time ago due to electron [Caprine](https://github.com/sindresorhus/caprine) tray icon is missing in system tray. After ticket reading it seems that's caused by missing SystemNotifier tray icon rendering feature in polybar.

## Why does polybar need this feature?
It is possible that more apps will move tray icon rendering to SystemNotifier implementation, thus polybar won't be able to render them.

## Describe the solution you'd like
Polybar supports SystemNotifier icons in tray rendering.

## Additional context
https://github.com/vladimiry/ElectronMail/issues/250#issuecomment-585541289

**Comment by vladimiry** [comment]
Electron v8 comes with Chromium 80.x which switched to use of StatusNotifier/DBus tray icon, [see](https://github.com/electron/electron/issues/21445#issuecomment-564488600):

> Possibly related chromium's [issue](https://bugs.chromium.org/p/chromium/issues/detail?id=419673) and [commit](https://github.com/chromium/chromium/commit/92171fce24a65a81f90705f974ace3f3dd92c349) (since 79.0.3923.0 tag/version).

If needed see some StatusNotifier tray icon support implementation in the following projects:
- https://git.xfce.org/panel-plugins/xfce4-statusnotifier-plugin/
- https://github.com/ubuntu/gnome-shell-extension-appindicator projects as an implementation example.

**Comment by patrick96** [comment]
Having StatusNotifierItem support has been on our TODO list for the longest time (see #913). However, we kind of always tied it to having first-class image support in polybar (#1088). I think that was a little too much at once. Just supporting StatusNotifierItem in the tray we currently have would be a better first step.

The xfce4-statusnotifier-plugin repo you linked does have a few thousand lines of code, so this won't be a quick thing for sure.

I will start reading up more on the spec. In the meantime, if anyone has experience in that area, I'd love your insight. Especially around what is actually needed for implementing this.

In #913 @NBonaparte mentioned:

> Menus are supposed to be implemented by the manager itself, rather than the clients (using dbusmenu). If we were using a toolkit like GTK or QT it would be easy, but since we're not, we would have to implement it ourselves.

This seems like a pretty big hurdle, but I could also be wrong since I have no idea how involved dealing with dbus will be.

**Comment by vladimiry** [comment]
> The xfce4-statusnotifier-plugin repo you linked does have a few thousand lines of code, so this won't be a quick thing for sure.

All I've found so far.

**Comment by vladimiry** [comment]
> The xfce4-statusnotifier-plugin repo you linked does have a few thousand lines of code, so this won't be a quick thing for sure.

https://github.com/i3/i3/issues/2088#issuecomment-378264464

**Comment by xeriab** [comment]
Hmm, I tried to figure out the whole problem. and if I'm right I need to create a workaround to support `statusnotifier` while keeping the `xembed` interface which not an easy thing to do! `statusnotifier` uses `GObject` and polybar has nothing to do with GTK/GObject stuff!

The best I can tell is we need to check how we can implement a `statusnotifier` like interface that uses/implements `xembed` interface.

**Comment by patrick96** [comment]
@xeriab This is indeed a hard problem and I am also not sure which way is the best way.

I have also had some ideas:

* Only support `statusnotifier` and recommend users use `xembed-sni-proxy` if they want to display xembed tray clients.
* Create a common abstraction for a "tray icon" and only use that for managing and displaying tray icons. This abstraction would be an interface on its own and completely independent from statusnotifier or xembed. We could then create two implementations for this interface: one for statusnotifier and one for xembed that allows us to handle both types of tray icons.

The first approach seems a little easier, but the benefit of the second approach is that it makes the entire tray more modular and we can more or less easily add support for new tray specifications and are not bound by either the xembed or statusnotifier spec.

I think a first draft implementation of a tray module would be very simple:

* Only displays tray client names, no icons
* Only supports statusnotifier

I think this is the biggest hurdle to get the ball rolling. It is improbable that we ever get a complete tray module with icons and everything in one go. This has to be incremental, icons and support for xembed can be added later.

**Comment by fauust** [comment]
Hi!
Here is dirty workaround that may work for some of you. My keybase and zulip tray icon reappeared but it may be a bit unstable and not so clean (see the following screenshot):

![screenshot_20200423_111131](https://user-images.githubusercontent.com/2197555/80081473-4551f400-8553-11ea-86d5-71a31f87ff6d.png)

Build and start `snixembed` (https://git.sr.ht/~steef/snixembed).
On debian 10 here are the steps:
```console
❯ sudo apt install libdbusmenu-gtk3-dev valac # you may have to install other deps, see README.md
❯ git clone https://git.sr.ht/~steef/snixembed && cd snixembed
❯ make
❯ ./snixembed --fork 
```
**Note:** I am using the master branch version (767f1b0) because with 0.2.0 I lost bluetooth and redshift icons.

Thanks @akashnimare (https://github.com/zulip/zulip-desktop/issues/936#issuecomment-618219556) for pointing me the snixembed project.

**Comment by Kryszak** [comment]
@fauust this made electron based systray icons visible again, thanks a lot!

**Comment by zonotope** [comment]
yep, this works for me too! yes, it is a hack, but it's a good stopgap until we can get full, first class sni support in polybar.

**Comment by 0pLuS0** [comment]
Is there any hope polybar will get this support?

I just found out nicotine+ doesn't work;

https://github.com/nicotine-plus/nicotine-plus/issues/3553

I tried snixembed and it doesn't work for me, apps that don't need it, when I launch them, that have system tray icons, they lockup and crash, so I can't use it.

We really need the support in polybar, but I'm assuming the request made 5 years ago, this isn't looking good.

THANKS
